File: attic/go/message/message.go
package message

import (
	"crypto/ed25519"
	"fmt"
	"time"

	"github.com/corpo/qntm/crypto"
	"github.com/corpo/qntm/identity"
	"github.com/corpo/qntm/pkg/cbor"
	"github.com/corpo/qntm/pkg/types"
)

// Manager handles message creation, encryption, decryption, and verification
type Manager struct {
	suite       *crypto.QSP1Suite
	identityMgr *identity.Manager
}

// NewManager creates a new message manager
func NewManager() *Manager {
	return &Manager{
		suite:       crypto.NewQSP1Suite(),
		identityMgr: identity.NewManager(),
	}
}

// CreateMessage creates a new encrypted message envelope
func (m *Manager) CreateMessage(
	senderIdentity *types.Identity,
	conversation *types.Conversation,
	bodyType string,
	body []byte,
	refs []interface{},
	ttlSeconds int64,
) (*types.OuterEnvelope, error) {
	if err := m.identityMgr.ValidateIdentity(senderIdentity); err != nil {
		return nil, fmt.Errorf("invalid sender identity: %w", err)
	}

	// Generate message ID
	msgID, err := m.identityMgr.GenerateMessageID()
	if err != nil {
		return nil, fmt.Errorf("failed to generate message ID: %w", err)
	}

	now := time.Now().Unix()
	expiryTS := now + ttlSeconds

	// Create inner payload structure for signing
	bodyStruct := struct {
		BodyType string        `cbor:"body_type"`
		Body     []byte        `cbor:"body"`
		Refs     []interface{} `cbor:"refs,omitempty"`
	}{
		BodyType: bodyType,
		Body:     body,
		Refs:     refs,
	}

	bodyStructBytes, err := cbor.MarshalCanonical(bodyStruct)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal body struct: %w", err)
	}

	bodyHash := m.suite.Hash(bodyStructBytes)

	// Create signable structure
	signable := types.Signable{
		Proto:     crypto.ProtoPrefix,
		Suite:     types.DefaultSuite,
		ConvID:    conversation.ID,
		MsgID:     msgID,
		CreatedTS: now,
		ExpiryTS:  expiryTS,
		SenderKID: senderIdentity.KeyID,
		BodyHash:  bodyHash,
	}

	// Sign the message
	signableBytes, err := cbor.MarshalCanonical(signable)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal signable: %w", err)
	}

	signature, err := m.suite.Sign(senderIdentity.PrivateKey, signableBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to sign message: %w", err)
	}

	// Create inner payload
	innerPayload := types.InnerPayload{
		SenderIKPK: senderIdentity.PublicKey,
		SenderKID:  senderIdentity.KeyID,
		BodyType:   bodyType,
		Body:       body,
		Refs:       refs,
		SigAlg:     "Ed25519",
		Signature:  signature,
	}

	// Serialize inner payload for encryption
	innerPayloadBytes, err := cbor.MarshalCanonical(innerPayload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal inner payload: %w", err)
	}

	// Create AAD structure
	aadStruct := types.AADStruct{
		Version:   types.ProtocolVersion,
		Suite:     types.DefaultSuite,
		ConvID:    conversation.ID,
		MsgID:     msgID,
		CreatedTS: now,
		ExpiryTS:  expiryTS,
		ConvEpoch: conversation.CurrentEpoch,
	}

	aadBytes, err := cbor.MarshalCanonical(aadStruct)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal AAD: %w", err)
	}

	// Derive nonce for this message
	nonce, err := m.suite.DeriveNonce(conversation.Keys.NonceKey, msgID[:])
	if err != nil {
		return nil, fmt.Errorf("failed to derive nonce: %w", err)
	}

	// Encrypt inner payload
	ciphertext, err := m.suite.Encrypt(conversation.Keys.AEADKey, nonce, innerPayloadBytes, aadBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to encrypt inner payload: %w", err)
	}

	// Compute AAD hash for optimization
	aadHash := m.suite.Hash(aadBytes)

	// Create outer envelope
	envelope := &types.OuterEnvelope{
		Version:    types.ProtocolVersion,
		Suite:      types.DefaultSuite,
		ConvID:     conversation.ID,
		MsgID:      msgID,
		CreatedTS:  now,
		ExpiryTS:   expiryTS,
		ConvEpoch:  conversation.CurrentEpoch,
		Ciphertext: ciphertext,
		AADHash:    aadHash,
	}

	return envelope, nil
}

// DecryptMessage decrypts and verifies a message envelope
func (m *Manager) DecryptMessage(
	envelope *types.OuterEnvelope,
	conversation *types.Conversation,
) (*types.Message, error) {
	if err := m.ValidateEnvelope(envelope); err != nil {
		return nil, fmt.Errorf("invalid envelope: %w", err)
	}

	// Check if message has expired
	if time.Now().Unix() > envelope.ExpiryTS {
		return nil, fmt.Errorf("message has expired")
	}

	// Check conversation ID matches
	if envelope.ConvID != conversation.ID {
		return nil, fmt.Errorf("conversation ID mismatch")
	}

	// Reconstruct AAD structure
	aadStruct := types.AADStruct{
		Version:   envelope.Version,
		Suite:     envelope.Suite,
		ConvID:    envelope.ConvID,
		MsgID:     envelope.MsgID,
		CreatedTS: envelope.CreatedTS,
		ExpiryTS:  envelope.ExpiryTS,
		ConvEpoch: envelope.ConvEpoch,
	}

	aadBytes, err := cbor.MarshalCanonical(aadStruct)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal AAD for decryption: %w", err)
	}

	// Verify AAD hash if present (optimization check)
	if len(envelope.AADHash) > 0 {
		computedAADHash := m.suite.Hash(aadBytes)
		if string(envelope.AADHash) != string(computedAADHash) {
			return nil, fmt.Errorf("AAD hash mismatch")
		}
	}

	// Derive nonce
	nonce, err := m.suite.DeriveNonce(conversation.Keys.NonceKey, envelope.MsgID[:])
	if err != nil {
		return nil, fmt.Errorf("failed to derive nonce for decryption: %w", err)
	}

	// Decrypt inner payload
	innerPayloadBytes, err := m.suite.Decrypt(conversation.Keys.AEADKey, nonce, envelope.Ciphertext, aadBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to decrypt message: %w", err)
	}

	// Deserialize inner payload
	var innerPayload types.InnerPayload
	if err := cbor.UnmarshalCanonical(innerPayloadBytes, &innerPayload); err != nil {
		return nil, fmt.Errorf("failed to unmarshal inner payload: %w", err)
	}

	if err := m.ValidateInnerPayload(&innerPayload); err != nil {
		return nil, fmt.Errorf("invalid inner payload: %w", err)
	}

	// Verify the signature
	verified, err := m.VerifyMessageSignature(envelope, &innerPayload)
	if err != nil {
		return nil, fmt.Errorf("signature verification failed: %w", err)
	}
	if !verified {
		return nil, fmt.Errorf("invalid message signature")
	}

	// Verify sender key ID matches public key
	if !m.identityMgr.VerifyKeyID(innerPayload.SenderIKPK, innerPayload.SenderKID) {
		return nil, fmt.Errorf("sender key ID does not match public key")
	}

	message := &types.Message{
		Envelope: envelope,
		Inner:    &innerPayload,
		Verified: verified,
	}

	return message, nil
}

// VerifyMessageSignature verifies the signature inside a decrypted message
func (m *Manager) VerifyMessageSignature(
	envelope *types.OuterEnvelope,
	innerPayload *types.InnerPayload,
) (bool, error) {
	// Reconstruct body structure for hashing
	bodyStruct := struct {
		BodyType string        `cbor:"body_type"`
		Body     []byte        `cbor:"body"`
		Refs     []interface{} `cbor:"refs,omitempty"`
	}{
		BodyType: innerPayload.BodyType,
		Body:     innerPayload.Body,
		Refs:     innerPayload.Refs,
	}

	bodyStructBytes, err := cbor.MarshalCanonical(bodyStruct)
	if err != nil {
		return false, fmt.Errorf("failed to marshal body struct for verification: %w", err)
	}

	bodyHash := m.suite.Hash(bodyStructBytes)

	// Reconstruct signable structure
	signable := types.Signable{
		Proto:     crypto.ProtoPrefix,
		Suite:     envelope.Suite,
		ConvID:    envelope.ConvID,
		MsgID:     envelope.MsgID,
		CreatedTS: envelope.CreatedTS,
		ExpiryTS:  envelope.ExpiryTS,
		SenderKID: innerPayload.SenderKID,
		BodyHash:  bodyHash,
	}

	signableBytes, err := cbor.MarshalCanonical(signable)
	if err != nil {
		return false, fmt.Errorf("failed to marshal signable for verification: %w", err)
	}

	// Verify signature
	err = m.suite.Verify(innerPayload.SenderIKPK, signableBytes, innerPayload.Signature)
	if err != nil {
		return false, nil // Signature verification failed, but no error in the process
	}

	return true, nil
}

// ValidateEnvelope validates the structure of an outer envelope
func (m *Manager) ValidateEnvelope(envelope *types.OuterEnvelope) error {
	if envelope == nil {
		return fmt.Errorf("envelope is nil")
	}

	// Check version
	if envelope.Version != types.ProtocolVersion {
		return fmt.Errorf("unsupported protocol version: %d", envelope.Version)
	}

	// Check suite
	if envelope.Suite != types.DefaultSuite {
		return fmt.Errorf("unsupported crypto suite: %s", envelope.Suite)
	}

	// Check timestamps
	if envelope.CreatedTS <= 0 {
		return fmt.Errorf("invalid created timestamp: %d", envelope.CreatedTS)
	}

	if envelope.ExpiryTS <= envelope.CreatedTS {
		return fmt.Errorf("expiry timestamp must be after created timestamp")
	}

	// Check ciphertext is not empty
	if len(envelope.Ciphertext) == 0 {
		return fmt.Errorf("ciphertext is empty")
	}

	// Check for future timestamp (allow some clock skew)
	maxFutureSkew := int64(600) // 10 minutes
	if envelope.CreatedTS > time.Now().Unix()+maxFutureSkew {
		return fmt.Errorf("message created timestamp is too far in the future")
	}

	return nil
}

// ValidateInnerPayload validates the structure of an inner payload
func (m *Manager) ValidateInnerPayload(inner *types.InnerPayload) error {
	if inner == nil {
		return fmt.Errorf("inner payload is nil")
	}

	// Check sender public key length
	if len(inner.SenderIKPK) != ed25519.PublicKeySize {
		return fmt.Errorf("invalid sender public key length: %d", len(inner.SenderIKPK))
	}

	// Check signature algorithm
	if inner.SigAlg != "Ed25519" {
		return fmt.Errorf("unsupported signature algorithm: %s", inner.SigAlg)
	}

	// Check signature length
	if len(inner.Signature) != ed25519.SignatureSize {
		return fmt.Errorf("invalid signature length: %d", len(inner.Signature))
	}

	// Check body type is not empty
	if inner.BodyType == "" {
		return fmt.Errorf("body type is empty")
	}

	return nil
}

// SerializeEnvelope serializes an envelope to canonical CBOR
func (m *Manager) SerializeEnvelope(envelope *types.OuterEnvelope) ([]byte, error) {
	return cbor.MarshalCanonical(envelope)
}

// DeserializeEnvelope deserializes an envelope from canonical CBOR
func (m *Manager) DeserializeEnvelope(data []byte) (*types.OuterEnvelope, error) {
	var envelope types.OuterEnvelope
	if err := cbor.UnmarshalCanonical(data, &envelope); err != nil {
		return nil, fmt.Errorf("failed to unmarshal envelope: %w", err)
	}

	if err := m.ValidateEnvelope(&envelope); err != nil {
		return nil, fmt.Errorf("invalid envelope: %w", err)
	}

	return &envelope, nil
}

// CheckExpiry checks if a message has expired
func (m *Manager) CheckExpiry(envelope *types.OuterEnvelope) bool {
	return time.Now().Unix() > envelope.ExpiryTS
}

// DefaultTTL returns a default TTL in seconds (30 days as per spec appendix)
func (m *Manager) DefaultTTL() int64 {
	return 30 * 24 * 60 * 60 // 30 days
}

// DefaultHandshakeTTL returns a default TTL for handshake messages (7 days as per spec appendix)
func (m *Manager) DefaultHandshakeTTL() int64 {
	return 7 * 24 * 60 * 60 // 7 days
}

File: python-dist/src/qntm/message.py
"""Message creation, encryption, decryption, and verification."""

import time

from .cbor import marshal_canonical, unmarshal
from .constants import (
    DEFAULT_HANDSHAKE_TTL_SECONDS,
    DEFAULT_SUITE,
    DEFAULT_TTL_SECONDS,
    PROTO_PREFIX,
    PROTOCOL_VERSION,
)
from .crypto import QSP1Suite
from .identity import generate_message_id, key_id_from_public_key, validate_identity

_suite = QSP1Suite()


def create_message(
    sender_identity: dict,
    conversation: dict,
    body_type: str,
    body: bytes,
    refs: list | None = None,
    ttl_seconds: int = DEFAULT_TTL_SECONDS,
    did: str | None = None,
) -> dict:
    """Create and encrypt a message, returning an outer envelope dict."""
    validate_identity(sender_identity)

    msg_id = generate_message_id()
    now = int(time.time())
    if not _safe_timestamp(ttl_seconds) or ttl_seconds <= 0 or not _safe_timestamp(now + ttl_seconds):
        raise ValueError("message TTL must be a positive safe integer within the timestamp range")
    expiry_ts = now + ttl_seconds

    # QSP v1.2 binds membership controls to their signed source epoch.
    if conversation.get('type') == 'group' and body_type in ('group_genesis','group_add','group_remove','group_rekey'):
        control = unmarshal(body)
        body = marshal_canonical({**control, 'group_epoch': conversation['currentEpoch']})

    # Build body structure for hashing
    body_struct = {"body": body, "body_type": body_type}
    if refs:
        body_struct["refs"] = refs

    body_struct_bytes = marshal_canonical(body_struct)
    body_hash = _suite.hash(body_struct_bytes)

    # Build signable
    signable = {
        "body_hash": body_hash,
        "conv_id": conversation["id"],
        "created_ts": now,
        "expiry_ts": expiry_ts,
        "msg_id": msg_id,
        "proto": PROTO_PREFIX,
        "sender_kid": sender_identity["keyID"],
        "suite": DEFAULT_SUITE,
    }

    signable_bytes = marshal_canonical(signable)
    signature = _suite.sign(sender_identity["privateKey"], signable_bytes)

    # Build inner payload
    inner_payload = {
        "body": body,
        "body_type": body_type,
        "sender_ik_pk": sender_identity["publicKey"],
        "sender_kid": sender_identity["keyID"],
        "sig_alg": "Ed25519",
        "signature": signature,
    }
    if refs:
        inner_payload["refs"] = refs

    inner_payload_bytes = marshal_canonical(inner_payload)

    # Build AAD
    aad_struct = {
        "conv_epoch": conversation["currentEpoch"],
        "conv_id": conversation["id"],
        "created_ts": now,
        "expiry_ts": expiry_ts,
        "msg_id": msg_id,
        "suite": DEFAULT_SUITE,
        "v": PROTOCOL_VERSION,
    }

    aad_bytes = marshal_canonical(aad_struct)

    # Encrypt
    nonce = _suite.derive_nonce(conversation["keys"]["nonceKey"], msg_id)
    ciphertext = _suite.encrypt(
        conversation["keys"]["aeadKey"], nonce, inner_payload_bytes, aad_bytes
    )
    aad_hash = _suite.hash(aad_bytes)

    envelope = {
        "aad_hash": aad_hash,
        "ciphertext": ciphertext,
        "conv_epoch": conversation["currentEpoch"],
        "conv_id": conversation["id"],
        "created_ts": now,
        "expiry_ts": expiry_ts,
        "msg_id": msg_id,
        "suite": DEFAULT_SUITE,
        "v": PROTOCOL_VERSION,
    }

    # Optional DID field — identity-layer metadata for DID resolution.
    # Backwards compatible: receivers that don't understand DIDs ignore it.
    if did is not None:
        envelope["did"] = did

    return envelope


def decrypt_message(envelope: dict, conversation: dict, *, allow_expired: bool = False) -> dict:
    """Decrypt and authenticate an envelope, rejecting expired messages by default.

    ``allow_expired=True`` is for verifying saved history only, never live receive
    or action authorization. Encryption, signature and timestamp validation remain
    mandatory in either mode.
    """
    validate_envelope(envelope)
    if allow_expired is not True and check_expiry(envelope):
        raise ValueError("message has expired")

    conv_id = bytes(envelope["conv_id"])
    expected_conv_id = bytes(conversation["id"])
    if conv_id != expected_conv_id:
        raise ValueError("conversation ID mismatch")

    # Reconstruct AAD
    aad_struct = {
        "conv_epoch": envelope.get("conv_epoch", 0),
        "conv_id": envelope["conv_id"],
        "created_ts": envelope["created_ts"],
        "expiry_ts": envelope["expiry_ts"],
        "msg_id": envelope["msg_id"],
        "suite": envelope["suite"],
        "v": envelope["v"],
    }

    aad_bytes = marshal_canonical(aad_struct)

    # Verify AAD hash if present
    if envelope.get("aad_hash") and len(bytes(envelope["aad_hash"])) > 0:
        computed_aad_hash = _suite.hash(aad_bytes)
        if bytes(envelope["aad_hash"]) != computed_aad_hash:
            raise ValueError("AAD hash mismatch")

    # Decrypt
    msg_id = bytes(envelope["msg_id"])
    nonce = _suite.derive_nonce(conversation["keys"]["nonceKey"], msg_id)
    inner_bytes = _suite.decrypt(
        conversation["keys"]["aeadKey"], nonce, bytes(envelope["ciphertext"]), aad_bytes
    )

    inner = unmarshal(inner_bytes)
    validate_inner_payload(inner)

    # Verify signature
    verified = verify_message_signature(envelope, inner)
    if not verified:
        raise ValueError("invalid message signature")

    # Verify sender key ID
    sender_pk = bytes(inner["sender_ik_pk"])
    sender_kid = bytes(inner["sender_kid"])
    computed_kid = key_id_from_public_key(sender_pk)
    if sender_kid != computed_kid:
        raise ValueError("sender key ID does not match public key")

    return {
        "envelope": envelope,
        "inner": inner,
        "verified": verified,
    }


def verify_message_signature(envelope: dict, inner_payload: dict) -> bool:
    """Verify the inner signature against the envelope."""
    body_struct = {
        "body": inner_payload["body"],
        "body_type": inner_payload["body_type"],
    }
    if inner_payload.get("refs") and len(inner_payload["refs"]) > 0:
        body_struct["refs"] = inner_payload["refs"]

    body_struct_bytes = marshal_canonical(body_struct)
    body_hash = _suite.hash(body_struct_bytes)

    signable = {
        "body_hash": body_hash,
        "conv_id": envelope["conv_id"],
        "created_ts": envelope["created_ts"],
        "expiry_ts": envelope["expiry_ts"],
        "msg_id": envelope["msg_id"],
        "proto": PROTO_PREFIX,
        "sender_kid": inner_payload["sender_kid"],
        "suite": envelope["suite"],
    }

    signable_bytes = marshal_canonical(signable)
    return _suite.verify(
        bytes(inner_payload["sender_ik_pk"]),
        signable_bytes,
        bytes(inner_payload["signature"]),
    )


def validate_envelope(envelope: dict) -> None:
    if envelope["v"] != PROTOCOL_VERSION:
        raise ValueError(f"unsupported protocol version: {envelope['v']}")
    if envelope["suite"] != DEFAULT_SUITE:
        raise ValueError(f"unsupported crypto suite: {envelope['suite']}")
    if not _safe_timestamp(envelope["created_ts"]) or envelope["created_ts"] <= 0:
        raise ValueError(f"invalid created timestamp: {envelope['created_ts']}")
    if not _safe_timestamp(envelope["expiry_ts"]) or envelope["expiry_ts"] <= envelope["created_ts"]:
        raise ValueError("expiry timestamp must be after created timestamp")
    ct = envelope["ciphertext"]
    if isinstance(ct, memoryview):
        ct = bytes(ct)
    if len(ct) == 0:
        raise ValueError("ciphertext is empty")
    if envelope["created_ts"] > int(time.time()) + 600:
        raise ValueError("message created timestamp is too far in the future")


def _safe_timestamp(value: object) -> bool:
    # Match JavaScript's exact integer domain; bool is not a wire timestamp.
    return type(value) is int and abs(value) <= 2**53 - 1


def check_expiry(envelope: dict) -> bool:
    """True once the current Unix second exceeds the signed expiry timestamp."""
    return int(time.time()) > envelope["expiry_ts"]


def validate_inner_payload(inner: dict) -> None:
    pk = inner["sender_ik_pk"]
    if isinstance(pk, memoryview):
        pk = bytes(pk)
    if len(pk) != 32:
        raise ValueError(f"invalid sender public key length: {len(pk)}")
    if inner["sig_alg"] != "Ed25519":
        raise ValueError(f"unsupported signature algorithm: {inner['sig_alg']}")
    sig = inner["signature"]
    if isinstance(sig, memoryview):
        sig = bytes(sig)
    if len(sig) != 64:
        raise ValueError(f"invalid signature length: {len(sig)}")
    if not inner.get("body_type"):
        raise ValueError("body type is empty")


def extract_did(envelope: dict) -> str | None:
    """Extract the optional DID URI from an envelope, if present."""
    return envelope.get("did")


def serialize_envelope(envelope: dict) -> bytes:
    return marshal_canonical(envelope)


def deserialize_envelope(data: bytes) -> dict:
    envelope = unmarshal(data)
    validate_envelope(envelope)
    return envelope


def default_ttl() -> int:
    return DEFAULT_TTL_SECONDS


def default_handshake_ttl() -> int:
    return DEFAULT_HANDSHAKE_TTL_SECONDS
